ARC: add ARCv3 64 bit support

evgeniy-paltsev opened the pull request:
Add new ARCv3 ISA support and add nSIM HS6x, QEMU HS6x simulation boards support.

evgeniy-paltsev committed:
ARC: make variables with regs and addresses bit agnostic

Make variables where we store CPU registers values and
memory addresses bit agnostic.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: prepare for building for ARCv3 HS6x

Do basic preparations for building code for ARCv3 HS6x
* add ISA_ARCV3 and CPU_HS6X config options
* add off_t type support for __ARC64__
* use elf64-littlearc format for linking
* use arc64 mcpu for CPU_HS6X

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: gcc-m-cpu: check for exact HS CPU time instead of whole family

As of today we check against CPU_ARCHS config option which may be
related to different HS CPUs (i.e. ARCv2 HS, ARCv3 HS, etc...).
Let's check against exact HS CPU type (CPU_HS3X) as we do EM
processors.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: ARCv3: add nsim_hs6x board

Add nSIM-based (simulator) board with single core ARCv3 HS6x 64 bit
CPU.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: split sys-io for common and ARCv2-only parts

Split ARC sys-io.h implementation for two part:
 * AUX reg accessors - ASM based which and valid for ARCv2
 * MMIO accessors - writen in C and valid for both ARCv2 and ARCv3

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: reuse headers for both ARCv3 and ARCv3 if possible

Reuse ARCv2 headers [where it is possible] for ARCv3.
In this commit we simply allow to use them for ARCv3, we'll
move it to proper folder and rename them [where it is required]
in the upcoming cleanup patch.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: rewrite ASM code with asm-compat macroses

Rewrite ARC assembler code with asm-compat macroses, so the same
code can be used for both ARCv2 (GNU and MWDT assemblers) and
ARCv3 (GNU assembler)

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: allow to build code for processors without ZOL

ARCv3 64 bit processors doesn't have Zero Delay Loop
(also named Zero Overhead Loop, ZOL) mechanism. Add kconfig
option to remove ZOL register save/restore so the code
can be build for both ARCv2 and ARCv3.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: make vector table bit agnostic

ARCv2 32 bit and ARCv3 64 bit share the same vector table
structure but with different vector entry size (32 and 64 bit),
so we can easily make vector table bit agnostic.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: boost default stacks in case of 64BIT

Increase stacks required for ARCv3 64-bit CI to pass.
The CMSIS stacks are for programs in samples/portability

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: workaround bloated structure access in ASM with _st_huge_offset

When we accessing bloated structure member we can exceed u9 operand
in store instruction. So we can use _st32_huge_offset macro instead
for 32 bit accesses

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: arch.h: setup ARCH_STACK_PTR_ALIGN correctly for 64 bit

Stack is aligned by 8 for 64 bit ISA and 4 for 32 bit ISA.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: mark accesses which are 32 bit despite of platform bittnes

Mark the places where we intentionally use st instead of STR for
code common for ARCv2 and ARCv3.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: ARCv3: add qemu HS6x board

Add QEMU board with single core ARCv3 HS6x 64 bit CPU

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

vineetgarc committed:
ARC: add asm-compat macros

Add asm-compat macroses to be able to co create assembler code
which works and can be build for both ARCv2 and ARCv3 ISAs

Signed-off-by: Vineet Gupta <vgupta@synopsys.com>
Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: add asm-compat macro for MWDT toolchain

Extend asm-compat macroses so they can be used with ARC MWDT
toolchain for ARCv2

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev commented:
Requires SDK with ARC64 support, being discussed here zephyrproject-rtos/sdk-ng#327, the buildkite job fails due to the lack of  ARCv3 toolchain in current Zephyr SDK version, so it can't build for nsim_hs6x board we have added.

evgeniy-paltsev commented:
The compliance check fails on the headers files with assembler macros, and it's expected to have incompatible syntax. So it should be considered as a false-positive.

dcpleung commented on arch/arc/Kconfig:
```diff
@@ -104,6 +105,18 @@ config FP_FPU_DA
 
 menu "ARC CPU Options"
 
+config ARC_LACKS_ZOL
```
Wonder if it would be better as `ARC_HAS_ZOL`. It should be conceptually more inline with others like `ARCH_HAS_*`.

evgeniy-paltsev replied:
Yep, make sense. Fixed for now, thanks.

evgeniy-paltsev commented:
The most interesting is the license check failure. As of today for ARCv3 we have asm-compat macros files which we share between the OSS projects so they are dual-licensed:
```
/* SPDX-License-Identifier: GPL-2.0 OR Apache-2.0 */
```

However the licence checker argues about that files, i.e.: `asm-macro-32-bit-gnu.h is not apache-2.0 licensed`.
I expect that it is just a false-positive. I haven't found any info about usage of dual-licensed code in Zephyr documentation, so @nashif , @ruuddw probably you can share any thought about it (or whom should I discuss it with)?

abrodkin commented on arch/arc/include/swap_macros.h:
```diff
@@ -520,6 +527,19 @@
 
+/*
+ * When we accessing bloated struct member we can exceed u9 operand in store
+ * instruction. So we nac use _st32_huge_offset macro instead
```
nac -> can

abrodkin commented on boards/arc/nsim/nsim_hs6x.yaml:
```diff
@@ -0,0 +1,12 @@
+simulation: nsim
+arch: arc
+toolchain:
+  - zephyr
```
Could you please add cross-toolchain here so that we may run Twister with cross-tools until ARC64 support is not merged in the Zephyr SDK?

evgeniy-paltsev replied:
Done, thanks!

abrodkin commented on boards/arc/nsim/support/nsim_hs6x.props:
```diff
@@ -0,0 +1,31 @@
+	nsim_isa_addr_size=64
+	nsim_isa_pc_size=64
+	nsim_isa_vec64=1
+	dcache=16384,64,2,a
```
Cache sizes seem to not match between nSIM & MDB

evgeniy-paltsev replied:
Done, thanks!

abrodkin commented on boards/arc/nsim/support/nsim_hs6x.props:
```diff
@@ -0,0 +1,31 @@
+	nsim_isa_mpy64=1
+	nsim_isa_mpy_option=9
+	nsim_hostlink=1
+	gverbose=0
```
I guess in Zephyr we don't need either `nsim_hostlink` or `gverbose` so let's dump it.

evgeniy-paltsev replied:
Done, thanks!

abrodkin commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -2,13 +2,20 @@
+  zephyr_compile_options(-mcpu=${GCC_M_CPU})
+  if(CONFIG_64BIT)
+    # ISA_ARCV3 & 64BIT
+    zephyr_compile_options(-ffixed-r13)
```
Do we really need that limitation in place?

evgeniy-paltsev replied:
Fixed, thanks!

abrodkin commented on soc/arc/snps_nsim/Kconfig.defconfig.hs6x:
```diff
@@ -0,0 +1,28 @@
+
+config NUM_IRQ_PRIO_LEVELS
+	# This processor supports 16 priority levels:
+	# 0 for Fast Interrupts (FIRQs) and 1-15 for Regular Interrupts (IRQs).
```
We don't currently support FIRQ in ARCv3 Zephyr port, so let's get rid of this comment.

evgeniy-paltsev replied:
Done, thanks!

abrodkin commented on soc/arc/snps_nsim/Kconfig.defconfig.hs6x:
```diff
@@ -0,0 +1,28 @@
+config CACHE_MANAGEMENT
+	default y
+
+
```
Extra newline?

abrodkin commented on soc/arc/snps_qemu/CMakeLists.txt:
```diff
@@ -1,2 +1,8 @@
+zephyr_compile_options(-mcpu=${GCC_M_CPU})
+
+if(CONFIG_64BIT)
+  zephyr_compile_options(-ffixed-r13)
```
Ditto, remove.

abrodkin reviewed (approved):
With a couple of tiny comments LGTM.

ruuddw commented on arch/arc/Kconfig:
```diff
@@ -83,11 +88,35 @@ config CPU_HS3X
+	  - LPcc instruction
+	  - LP_COUNT core reg
+	  - LP_START, LP_END aux regs
+	  This optional removes any use of ZOL instructions/regs from code
```
optional -> option

evgeniy-paltsev replied:
Fixed, thanks!

ruuddw commented on arch/arc/Kconfig:
```diff
@@ -147,6 +177,7 @@ config ARC_FIRQ_STACK_SIZE
 
 config ARC_HAS_STACK_CHECKING
 	bool "ARC has STACK_CHECKING"
+	depends on ISA_ARCV2
```
No stack checking planned at all for V3, or no support in current implementation(s)?
For arch level kconfigs, I think the 'depends' should reflect the ARCv3 ISA, not implementations.

ruuddw replied:
I didn't check all, but maybe there are others more relevant to add a 'depend', e.g. HAS_SECURE.

evgeniy-paltsev replied:
> I didn't check all, but maybe there are others more relevant to add a 'depend', e.g. HAS_SECURE.

Yep, HAS_SECURE dependency was missing, now it's fixed, rhanks!

evgeniy-paltsev replied:
> No stack checking planned at all for V3, or no support in current implementation(s)?

No support in current implementation

> For arch level kconfigs, I think the 'depends' should reflect the ARCv3 ISA, not implementations.

I tend to disagree here. We describe the Zephyr configuration via Kconfig (and not a ISA or exact CPU documentation). We shouldn't allow to user to enable feature combination we don't support - it will be confusing and not user-friendly.

ruuddw replied:
I guess what triggered my remark was more that if the hardware supports it, and Zephyr supports it for other ARC variants, we should support it for ARCV3 as well. So it depends on timing: in this PR it is not supported, in a next one it probably will.

ruuddw reviewed (approved):
Few CI issues to be still resolved, but for the rest looks good.

evgeniy-paltsev committed:
ARC: add TLS support for ARCv3

For ARCv3 the register is fixed to r30, so we don't need to
configure it at compile-time.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

tejlmand commented on arch/arc/Kconfig:
```diff
@@ -104,6 +104,18 @@ config FP_FPU_DA
 menu "ARC CPU Options"
 
+config ARC_HAS_ZOL
+	bool "Use Zero Delay hardware loops"
```
Should this really be user configurable.
Sounds a bit like something that should be promptless and selected based on soc/arc.

evgeniy-paltsev replied:
Sounds reasonable, fixed for now. Thanks!

galak commented:
Can you update this PR and remove the boards?  This way we should be able to merge the main changes and not run into CI since we don't have a full SDK yet.

galak commented:
> To explain further, we don't have any means today to have different boards utilize different SDK versions. So we can't have the ARC64 boards utilize the 0.13.0-alpha-1 and all other zephyr platforms utilize 0.12.4. So figure we can at least merge the non-board code for now and add the boards once we have either a solution to running different SDK versions or the 0.13.0 SDK release is out that supports everything.

(find me on Slack if you need more details).

galak commented:
> If you get that to work I'm fine with that. Just make sure that we do that as its own commit. Can you add that to this PR, while we have the other PR going for testing purposes.

Actually can we just not have `zephyr` listed in boards/*/foo.yaml under toolchains?

galak commented on boards/arc/qemu_arc/qemu_arc_hs6x.yaml:
```diff
@@ -0,0 +1,12 @@
+simulation: qemu
+arch: arc
+toolchain:
+  - zephyr
```
can we remove `zephyr` for now.

galak commented:
Not sure if the changes to `boards/arc/qemu_arc/board.cmake` will cause an issue.

evgeniy-paltsev commented:
> Not sure if the changes to boards/arc/qemu_arc/board.cmake will cause an issue.

I've tested this locally with both SDKs (for ARCv2) and 0.13.0-alpha-1 (for ARCv3) - everything works fine. We don't specify the board as we are fine with default value.

evgeniy-paltsev commented on boards/arc/nsim/nsim_hs6x.yaml:
```diff
@@ -0,0 +1,13 @@
+simulation: nsim
+arch: arc
+toolchain:
+  - zephyr
```
That excludes platform from Zephyr CI. I've changed it for this PR, thanks!

evgeniy-paltsev commented:
To sum:
 * The ARCv3 64 platforms can be built and tested with 0.13.0-alpha-1 SDK, the results can be found here: https://github.com/zephyrproject-rtos/zephyr/pull/34961
 * We don't enable testing for ARCv3 64 by default in CI as we keep 0.12.4 SDK as default
 * The compliance check fails on the headers files with assembler macros, and it's expected to have incompatible syntax. So it should be considered as a false-positive.
